internal/component/mimir/alerts/kubernetes/alerts.go

```diff
@@ -0,0 +1,370 @@
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+
+	promExternalVersions "github.com/prometheus-operator/prometheus-operator/pkg/client/informers/externalversions"
+	promListers_v1alpha1 "github.com/prometheus-operator/prometheus-operator/pkg/client/listers/monitoring/v1alpha1"
+	promVersioned "github.com/prometheus-operator/prometheus-operator/pkg/client/versioned"
+)
+
+const (
+	configurationUpdate = "configuration-update"
+)
+
+var (
+	errShutdown = errors.New("component is shutting down")
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:      "mimir.alerts.kubernetes",
+		Stability: featuregate.StabilityExperimental,
+		Args:      Arguments{},
+		Exports:   nil,
+		Build: func(o component.Options, c component.Arguments) (component.Component, error) {
+			return New(o, c.(Arguments))
+		},
+	})
+}
+
+type Component struct {
+	log  log.Logger
+	opts component.Options
+	args Arguments
+
+	// The client via which Alertmanager configs are sent to Mimir.
+	mimirClient mimirClient.AlertmanagerInterface
+
+	// Signal an update to the Arguments.
+	configUpdates chan ConfigUpdate
+
+	ticker *time.Ticker
+
+	healthMut sync.RWMutex
+	health    component.Health
+	metrics   *metrics
+
+	// Connection to the Kubernetes API.
+	k8sClient kubernetes.Interface
+	// Selector for "Namespace" k8s resources which to watch.
+	namespaceSelector labels.Selector
+	// Selector for "AlertmanagerConfig" k8s resources which to watch.
+	cfgSelector labels.Selector
+	// A Prometheus Operator client via which "AlertmanagerConfigs" are retrieved from the Kubernetes API.
+	promClient promVersioned.Interface
+	// The event processor that watches for changes in the Kubernetes API and updates the Mimir Alertmanager configs.
+	eventProcessor *eventProcessor
+}
+
+type ConfigUpdate struct {
+	args Arguments
+}
+
+var _ component.Component = (*Component)(nil)
+var _ component.HealthComponent = (*Component)(nil)
+
+// TODO: Implement DebugInfo()
+// var _ component.DebugComponent = (*Component)(nil)
+
+// New creates a new Component and initializes required clients based on the provided configuration.
+// TODO: Add clustering support
+func New(o component.Options, args Arguments) (*Component, error) {
+	c, err := newNoInit(o, args)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.init()
+	if err != nil {
+		return nil, fmt.Errorf("initializing component failed: %w", err)
+	}
+
+	return c, nil
+}
+
+func newNoInit(o component.Options, args Arguments) (*Component, error) {
+	m := newMetrics()
+	if err := m.register(o.Registerer); err != nil {
+		return nil, fmt.Errorf("registering metrics failed: %w", err)
+	}
+
+	c := &Component{
+		log:           o.Logger,
+		opts:          o,
+		args:          args,
+		configUpdates: make(chan ConfigUpdate),
+		metrics:       m,
+		ticker:        time.NewTicker(args.SyncInterval),
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	c.startupWithRetries(ctx, c, c)
```

Comment:
Could this block forever?

If you have a bad config this would never start because we are doing infinite retries. We would also not get any args updates that would fix the issue because we are blocked here right?